plugin/metrics/prometheus/metricsstore/reader.go

```diff
@@ -16,22 +16,45 @@ package metricsstore
 
 import (
 	"context"
+	"fmt"
+	"net"
 	"time"
 
 	"go.uber.org/zap"
 
+	"github.com/jaegertracing/jaeger/pkg/multierror"
 	"github.com/jaegertracing/jaeger/proto-gen/api_v2/metrics"
 	"github.com/jaegertracing/jaeger/storage/metricsstore"
 )
 
 // MetricsReader is a Prometheus metrics reader.
-type MetricsReader struct{}
+type MetricsReader struct {
+	url    string
+	logger *zap.Logger
+}
 
 // NewMetricsReader returns a new MetricsReader, assigning the first reachable host:port from the provided list.
 // This host:port forms part of the URL to call when making queries to the underlying metrics store.
 func NewMetricsReader(logger *zap.Logger, hostPorts []string, connTimeout time.Duration) (*MetricsReader, error) {
-	// TODO: Implement me
-	return &MetricsReader{}, nil
+	if len(hostPorts) < 1 {
+		return nil, fmt.Errorf("no prometheus query host:port provided")
+	}
+	errs := make([]error, 0)
+	for _, hostPort := range hostPorts {
```

Comment:
This is kind of an odd behavior, why do we need to loop through hosts looking for a reachable one? Why wouldn't they all be reachable? Or to put another way, what's the point of providing unreachable host:port pairs as input?

If these addresses are somehow "unstable", then they might as well become unreachable _after_ the reader is created.

Reply:
Thanks for pointing this out @yurishkuro.

The initial motivation was to maintain config consistency with ES and Cassandra span stores in how they both supply multiple server URLs, presumably for redundancy but yes, good call that the selected address could be unreachable after the reader is created.

This constructor behaviour was based on following ES's lead in how it initiates a new client by health checking each server: https://github.com/olivere/elastic/blob/v6.2.35/client.go#L1153.

However, the piece I'm missing which you correctly picked up on, is maintaining this pool of connections and finding the first healthy connection on each request, for example: https://github.com/olivere/elastic/blob/v6.2.35/client.go#L1301

The trade-off here is that it's a fair bit more complicated.

What direction do you suggest we take? Have a simple single host:port configuration? Or support a pool of connections that we select on each request?

Reply:
Does Prom only expose HTTP API, or is there gRPC version? gRPC would've taken care of the connection pooling.

Reply:
To the best of my knowledge, Prometheus only exposes an [HTTP API](https://prometheus.io/docs/prometheus/latest/querying/api/). 

I found some notes in Prometheus docs referring to [using gRPC for remote storage reads/writes in the future](https://prometheus.io/docs/prometheus/latest/storage/#overview), but that's not related to the query API.

Reply:
Another alternative is to use the Prometheus go client; but it means including another dependency.

Worth noting is this client takes a single URL: https://github.com/prometheus/client_golang/blob/master/api/client.go#L78

Given we're only using in a single Prometheus endpoint, I feel the Prometheus client offers more than we need and may be a bit of an overkill.
 
At the same time, we can adopt this client's approach of just a single connection URL, which I assume would be the more idiomatic given it's an official Prometheus client.

Reply:
aren't we already using prom client for metrics?

Reply:
Yes, you're right. However, the [README](https://github.com/prometheus/client_golang#client-for-the-prometheus-http-api) says that the client used to query the HTTP API is still in alpha:

> The api/prometheus directory contains the client for the Prometheus HTTP API. It allows you to write Go applications that query time series data from a Prometheus server. It is still in alpha stage.

Reply:
If we want to proceed with the Prom client library, I think it would be best we modify our [metrics data model](https://github.com/jaegertracing/jaeger/tree/master/model/proto/metrics) to match prometheus' to avoid mapping between OpenTelemetry and Prometheus' data models. I guess this might mean using [OpenMetrics](https://github.com/OpenObservability/OpenMetrics/blob/main/proto/openmetrics_data_model.proto), unless if there's a good reason to stick with OpenTelemetry's.

Otherwise, should we maintain the simple http client as we have now, configured with a single Prom address?

Reply:
I don't have a strong opinion. Which of the two formats more mature at this point? I thought OTLP was still changing its proto.

Reply:
We just discussed this in the Jaeger call. General agreement was that OpenMetrics would be a better choice:
  * OTLP is still changing, and we already took a copy of the data model which might be different now
  * OTLP is supposed to be fully compatible with OpenMetrics by the end of the year
  * OpenMetrics is the de-factor Prometheus format used by many different backends already
  * OpenMetrics is stable
  * OTEL will support OTLP<->OpenMetrics conversion in the future so we would still be able to implement support for OTLP-native backends

Reply:
Okay, thanks for raising this in the Jaeger call. Those reasons make sense so I'll make the switch over to OpenMetrics.

It also sounds like we're okay with using the Prom client for HTTP queries (even though it's in alpha), but please correct me if my assumption is wrong.

Reply:
> It also sounds like we're okay with using the Prom client for HTTP queries

sgtm